core/src/main/java/org/kiwix/kiwixmobile/core/webserver/ZimHostActivity.kt

```diff
@@ -0,0 +1,290 @@
+
+import android.app.ProgressDialog
+import android.content.ComponentName
+import android.content.Context
+import android.content.Intent
+import android.content.ServiceConnection
+import android.os.Bundle
+import android.os.IBinder
+import android.util.Log
+import android.widget.Toast
+import kotlinx.android.synthetic.main.activity_zim_host.recyclerViewZimHost
+import kotlinx.android.synthetic.main.activity_zim_host.serverTextView
+import kotlinx.android.synthetic.main.activity_zim_host.startServerButton
+import kotlinx.android.synthetic.main.activity_zim_host.toolbar
+import org.kiwix.kiwixmobile.core.BuildConfig
+import org.kiwix.kiwixmobile.core.R
+import org.kiwix.kiwixmobile.core.base.BaseActivity
+import org.kiwix.kiwixmobile.core.utils.AlertDialogShower
+import org.kiwix.kiwixmobile.core.utils.KiwixDialog
+import org.kiwix.kiwixmobile.core.utils.ServerUtils
+import org.kiwix.kiwixmobile.core.wifi_hotspot.HotspotService
+import org.kiwix.kiwixmobile.core.wifi_hotspot.HotspotService.ACTION_CHECK_IP_ADDRESS
+import org.kiwix.kiwixmobile.core.wifi_hotspot.HotspotService.ACTION_START_SERVER
+import org.kiwix.kiwixmobile.core.wifi_hotspot.HotspotService.ACTION_STOP_SERVER
+import org.kiwix.kiwixmobile.core.zim_manager.fileselect_view.SelectionMode
+import org.kiwix.kiwixmobile.core.zim_manager.fileselect_view.adapter.BookOnDiskDelegate
+import org.kiwix.kiwixmobile.core.zim_manager.fileselect_view.adapter.BooksOnDiskAdapter
+import org.kiwix.kiwixmobile.core.zim_manager.fileselect_view.adapter.BooksOnDiskListItem
+import org.kiwix.kiwixmobile.core.zim_manager.fileselect_view.adapter.BooksOnDiskListItem.BookOnDisk
+import java.util.ArrayList
+import javax.inject.Inject
+
+class ZimHostActivity : BaseActivity(), ZimHostCallbacks, ZimHostContract.View {
+  @Inject
+  internal lateinit var presenter: ZimHostContract.Presenter
+  @Inject
+  internal lateinit var alertDialogShower: AlertDialogShower
+  private lateinit var booksAdapter: BooksOnDiskAdapter
+  private lateinit var bookDelegate: BookOnDiskDelegate.BookDelegate
+  private var hotspotService: HotspotService? = null
+  private var ip: String? = null
+  private var serviceConnection: ServiceConnection? = null
+  private var progressDialog: ProgressDialog? = null
+  private val tag = "ZimHostActivity"
+  private val ipStateKey = "ip_state_key"
+
+  private val selectedBooksPath: ArrayList<String>
+    get() {
+      return booksAdapter.items
+        .filter(BooksOnDiskListItem::isSelected)
+        .filterIsInstance<BookOnDisk>()
+        .map {
+          if (BuildConfig.DEBUG) {
+            Log.v(tag, "ZIM PATH : " + it.file.absolutePath)
+          }
+          it.file.absolutePath
+        }
+        as ArrayList<String>
+    }
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_zim_host)
+
+    setUpToolbar()
+
+    bookDelegate = BookOnDiskDelegate.BookDelegate(sharedPreferenceUtil, null, null,
```

Comment:
If you look at the constructor of BookDelegate you will see that it has default arguments for its parameters, this means we can choose to omit them and the default values will be taken, in this case null.

You can start entering parameters as normal, then skip any default ones you want to skip and then if you want to set one of the default arguments refer to it by name eg `BookOnDiskDelegate.BookDelegate(sharedPreferenceUtil, multiSelectAction = { select(bookOnDiskItem) })`

Reply:
https://github.com/kiwix/kiwix-android/pull/1569/commits/6b918125101e56fe10b9392a5817f23c5bcac93c